topics/dev/tutorials/interactive-tools/tutorial.md

```diff
@@ -525,6 +530,535 @@ our new Docker container as a Galaxy tool.
+
+```sh
+$__tool_directory__
+$__root_dir__
+$__user_id__
+$__user_email__
+```
+
+It can also be useful to create and inject environment variables into the tool context. This can be acheived using the `<environment variables>` tag in the tool XML. The [RStudio GxIT](https://github.com/galaxyproject/galaxy/blob/b180b7909dc3fe2750fbc8b90214e201eb276794/tools/interactive/interactivetool_rstudio.xml#L12) again provides an example of this:
+
+```xml
+<environment_variables>
+    <environment_variable name="HISTORY_ID" strip="True">${__app__.security.encode_id($jupyter_notebook.history_id)}</environment_variable>
+    <environment_variable name="REMOTE_HOST">${__app__.config.galaxy_infrastructure_url}</environment_variable>
+    <environment_variable name="GALAXY_WEB_PORT">8080</environment_variable>
+    <environment_variable name="GALAXY_URL">$__galaxy_url__</environment_variable>
+    <environment_variable name="DEBUG">true</environment_variable>
+    <environment_variable name="DISABLE_AUTH">true</environment_variable>
+    <environment_variable name="API_KEY" inject="api_key" />
+</environment_variables>
+```
+
+## Galaxy history interaction
+We have demonstrated how to pass an input file to the Docker container. But what if the application needs to interact with the user's Galaxy history? For example, if the user creates a file within the application. That's where the environment variables created in the tool XML become useful.
+
+> <tip-title>Access histories in R</tip-title>
+> From the [R-Studio GxIT](https://github.com/galaxyproject/galaxy/blob/dev/tools/interactive/interactivetool_rstudio.xml) we can see that there is [an R library](https://github.com/hexylena/rGalaxyConnector) that allows us to interact with Galaxy histories.
+>
+> "The convenience functions `gx_put()` and `gx_get()` are available to you to interact with your current Galaxy history. You can save your workspace with `gx_save()`"
+>
+> Under the hood, this library uses [galaxy_ie_helpers](https://github.com/bgruening/galaxy_ie_helpers) - a Python interface to Galaxy histories written with [BioBlend](https://github.com/galaxyproject/bioblend). You could also use BioBlend directly (or even the Galaxy REST API) if your GxIT requires a more flexible interface than these wrappers provide.
+>
+{: .tip}
+
+
+</div>
+
+<div class="Jupyterlab" markdown="1">
+
+# The jupyterlab
+
+The jupyterlab environement that we will wrap in this tutorial is a simple jupyterlab tool which
+allows the user to upload data in an jupyter environment, manipulate notebooks, and download
+outputs. 
+
+Our example jupyterlab can already be found [online](https://github.com/usegalaxy-eu/docker-copernicus-notebooks) 
+But you can also check an other implementation of [jupyterlabs](https://github.com/bgruening/docker-jupyter-notebook/blob/master/Dockerfile).
+In the following sections, we will study how it can be built into a GxIT.
+
```

Comment:
```suggestion
The JupyterLab environment that we will wrap in this tutorial is a simple JupyterLab tool which
allows the user to upload data in a Jupyter environment, manipulate notebooks, and download
outputs. 

Our example JupyterLab can already be found [online](https://github.com/usegalaxy-eu/docker-copernicus-notebooks) 
But you can also check another implementation of [JupyterLabs](https://github.com/bgruening/docker-jupyter-notebook/blob/master/Dockerfile).
In the following sections, we will study how it can be built into a GxIT.

```